src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSource.java

```diff
@@ -1125,9 +1133,17 @@ public String getDisplayName() {
             return "Bitbucket";
         }
 
-        public FormValidation doCheckCredentialsId(@QueryParameter String value,
+        public FormValidation doCheckCredentialsId(@CheckForNull @AncestorInPath SCMSourceOwner context,
+                                                   @QueryParameter String value,
                                                    @QueryParameter String bitbucketServerUrl) {
             if (!value.isEmpty()) {
+                if (CredentialsMatchers.firstOrNull(
+                        CredentialsProvider.lookupCredentials(StandardCertificateCredentials.class,
+                                context,
+                                context instanceof Queue.Task ? Tasks.getDefaultAuthenticationOf((Queue.Task) context) : ACL.SYSTEM,
+                                URIRequirementBuilder.fromUri(bitbucketServerUrl).build()), CredentialsMatchers.withId(value)) != null) {
```

Comment:
You will need both `withId` and `AuthenticationTokens.matcher(BitbucketCredentials.class)` to handle the case where there are two different types of credentials with the same id at different layers in the folder hierarchy.

Reply:
The point being that the drop-down list will be showing only those that `AuthenticationTokens.matcher(BitbucketCredentials.class)` so may hide higher up ones that happen to have the same id but don't match the type we want...

say at the root there is a userpass with id `foo` and one folder down there is a certificate with id `foo`... if the url is `http` we should be able to select `foo` and mean the userpass...

and such a config could actually be desired, e.g. against `https` we want to use the stronger client cert... against `http` we have no choice... and we set up everything to always have id of `foo` so that the choice is easy... not a config I'd recommend... but it is logical... and i think credentials-api would support it